### PR TITLE
Ensure that the reloading of a config file only happenes once

### PR DESCRIPTION
**Pull request by jrouzierinverse**
## Description

Ensure that configuration files are only reloaded once to avoid reparsing
## Impacts

This impact all services
bin/pfcmd configreload
bin/pfcmd configreload hard
## Delete branch after merge

NO

### CODE REVIEW

File: addons/dev-helpers/config-reload-loop.pl [hunk 1]
```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/perl
+
+=head1 NAME
+
+loop add documentation
```
Comment (julsemaan) [inline addons/dev-helpers/config-reload-loop.pl]: Remove this or add documentation

File: addons/dev-helpers/config-reload-loop.pl [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    last unless keys %KIDS;
+}
+
+=head1 AUTHOR
+
+Inverse inc. <info@inverse.ca>
+
+=head1 COPYRIGHT
+
+Copyright (C) 2005-2014 Inverse inc.
```
Comment (julsemaan) [inline addons/dev-helpers/config-reload-loop.pl]: we're in 2015

File: addons/dev-helpers/config-reload-loop.pl [hunk 3]
```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/perl
+
+=head1 NAME
+
+loop add documentation
+
+=cut
+
+=head1 DESCRIPTION
```
Comment (julsemaan) [inline addons/dev-helpers/config-reload-loop.pl:9]: I think this file would deserve a short description of what it's doing.

File: lib/pf/FileLocker.pm [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    return F_WRLCK == $fs->l_type && $fs->l_pid != $$;
+}
+
+=head1 AUTHOR
+
+Inverse inc. <info@inverse.ca>
+
+=head1 COPYRIGHT
+
+Copyright (C) 2005-2014 Inverse inc.
```
Comment (julsemaan) [inline lib/pf/FileLocker.pm]: 2015

File: lib/pf/config/cached.pm [hunk 5]
```diff
@@ -672,36 +682,39 @@ sub _makeFileLock {
...
     my $logger = get_logger();
-    $logger->trace("ReadConfig for $file");
     my $new_self = $self->computeFromPath(
         $file,
         sub {
             #reread files
             #lock will release when out of scope
-            my $lock = lockFileForReading($file);
-            $result = $self->SUPER::ReadConfig();
+#            my $lock = lockFileForReading($file);
```
Comment (julsemaan) [inline lib/pf/config/cached.pm:706]: you should remove that line if it's not necessary anymore

File: lib/pf/config/cached.pm [hunk 6]
```diff
@@ -782,38 +794,109 @@ sub computeFromPath {
...
+sub GotReloadWriteLock {
+    my ($self) = @_;
+    my $locker = _lockFileForOnReload($self->GetFileName);
+    $locker->blocking(0); 
+    return $locker->writeLock;
+}
+
+=head2 LockFileHasChanged
+
+TODO: documention
```
Comment (julsemaan) [inline lib/pf/config/cached.pm]: let's do it then

File: lib/pf/config/cached.pm [hunk 7]
```diff
@@ -782,38 +794,109 @@ sub computeFromPath {
...
+    return $locker->isWriteLockedByAnother;
+}
+
+=head2 GotReloadWriteLock
+
+Gets an existing write lock
+
+=cut
+
+sub GotReloadWriteLock {
```
Comment (julsemaan) [inline lib/pf/config/cached.pm]: Why is it Got if it 'gets an existing write lock'

File: lib/pf/config/cached.pm [hunk 8]
```diff
@@ -948,16 +1042,19 @@ sub addPostReloadCallbacks {
...
 sub DESTROY {
     my ($self) = @_;
+    #Do nothing if no destroy is set
+    return if $NO_DESTROY || $self->{no_destroy};
+
     my $file = $self->GetFileName;
     return unless $file;
     foreach my $hash_ref (@ON_DESTROY_REFS) {
-        delete $hash_ref->{$file};
+        #delete $hash_ref->{$file};
```
Comment (julsemaan) [inline lib/pf/config/cached.pm]: if it's not useful anymore, delete the line

File: lib/pf/config/cached.pm [hunk 9]
```diff
@@ -1079,21 +1177,28 @@ sub clearAllConfigs {
...
-        my $fh;
-        open($fh,">$cache_control_file") or die "Can't create $cache_control_file\nPlease run 'pfcmd fixpermissions'";
-        close($fh);
-    }
-    if(-e $cache_control_file) {
-        sysopen(my $fh,$cache_control_file,O_RDWR | O_CREAT);
-        POSIX::2008::futimens(fileno $fh);
-        close($fh);
-    }
+sub _ExpireFile {
```
Comment (julsemaan) [inline lib/pf/config/cached.pm:1166]: pod doc for the sub ?

File: lib/pf/config/cached.pm [hunk 10]
```diff
@@ -1079,21 +1177,28 @@ sub clearAllConfigs {
...
+    close($fh);
     my (undef,undef,$uid,$gid) = getpwnam('pf');
-    chown($uid,$gid,$cache_control_file);
-    return 0;
+    chown($uid,$gid,$file);
+    umask($old_mask);
+}
+
+
+sub ExpireLockFile {
```
Comment (julsemaan) [inline lib/pf/config/cached.pm:1183]: doc ?

File: lib/pf/config/cached.pm [hunk 11]
```diff
@@ -1079,21 +1177,28 @@ sub clearAllConfigs {
...
+    my ($self) = @_;
+    my $locker = lockFileForWriting($self->GetFileName);
+    $self->_ExpireFile($self->GetLockFileName());
+}
+
+sub ExpireFile {
```
Comment (julsemaan) [inline lib/pf/config/cached.pm:1196]: doc ?

File: sbin/pfsetvlan [hunk 12]
```diff
@@ -2053,9 +2053,8 @@ sub getSwitch {
     my $lockLogger = Log::Log4perl->get_logger('pfsetvlan::locking');
     {
         lock $switchFactory_locker;
-    #    $pf::ConfigStore::SwitchOverlay::switches_overlay_cached_config->ReadConfig();
```
Comment (julsemaan) [inline sbin/pfsetvlan:2056]: Remove the line if it's not useful anymore

File: addons/dev-helpers/config-reload-loop.pl (hunk 2 above)
Comment (jrouzierinverse) [inline addons/dev-helpers/config-reload-loop.pl]: It was created in 2014

**Comment by jrouzierinverse** [comment]
Please review again

File: sbin/pfsetvlan (hunk 12 above)
Comment (julsemaan) [inline sbin/pfsetvlan:2056]: Is this line really useful ?

File: lib/pf/config/cached.pm (hunk 5 above)
Comment (julsemaan) [inline lib/pf/config/cached.pm:706]: Is this line really useful ?

File: sbin/pfsetvlan (hunk 12 above)
Comment (jrouzierinverse) [inline sbin/pfsetvlan:2056]: It is already removed

File: addons/dev-helpers/config-reload-loop.pl [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    onfilereload => sub {print STDERR "$$ onfilereload " . time . "\n"});
+$pf::ConfigStore::Switch::switches_cached_config->addCacheReloadCallbacks(
+    oncachereload => sub {print STDERR "$$ oncachereload " . time . "\n"});
+
+{
+    Cache::Memcached->disconnect_all;
+    pf::CHI->clear_memoized_cache_objects;
+    db_disconnect();
+}
+#    local $pf::config::cached::NO_DESTROY = 1;
```
Comment (julsemaan) [inline addons/dev-helpers/config-reload-loop.pl]: Remove that line. I'm not crazy now, that one is really there !

File: lib/pf/config/cached.pm [hunk 14]
```diff
@@ -627,35 +584,88 @@ Locks the lock file for writing a file
...
+
+=cut
+
+sub _makeLocker {
+    my ($file) = @_;
+    my $lockfile = _makeFileLock($file);
+    umask 0;
+    sysopen(my $fh,$lockfile,O_CREAT|O_RDWR,0660);
+#    my (undef,undef,$uid,$gid) = getpwnam('pf');
+#    chown($uid,$gid,$lockfile);
```
Comment (julsemaan) [inline lib/pf/config/cached.pm]: is that needed ?

**Comment by jrouzierinverse** [comment]
Please review

**Comment by extrafu** [comment]
Was it thoroughly tested before the merge? Ie., on a test environment with thousands of switches and some load during the tests (adding a switch, deleting one, etc.) ?

**Comment by julsemaan** [comment]
I didn't do it personally.

**Comment by extrafu** [comment]
Let's plan for this please!

**Comment by jrouzierinverse** [comment]
Update CHI to version 0.59 in order to test this.
Either grab it from CPAN or ask @fdurand to update our repo

**Comment by julsemaan** [comment]
I'll test it and report but usually pull requests should be fully tested before they are submitted 'for review'

**Comment by julsemaan** [comment]
Also the package CHI 0.59 should have been packaged and put in the devel repo before this was submitted for review

**Comment by julsemaan** [comment]
We're also missing perl-File-FcntLock.

I've unmerged the branch. Can you fix the packaging issues, make sure it will work in devel and open a new pull request for it

**Comment by extrafu** [comment]
This is exactly my "point" on why the pull request should be tested before being merged. To sum it up:
1. it should be tested before being submitted
2. it should be reviewed and tested
3. it should be merged

Code review brings a different perspective to the submitted changes. Testing before merging does the same.

Thanks!

**Pull request merged**